gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/RefreshTokenAuth2Provider.java

```diff
@@ -0,0 +1,162 @@
+package com.google.cloud.hadoop.fs.gcs;
+
+import static com.google.api.client.util.Preconditions.checkNotNull;
+
+import com.google.api.client.auth.oauth2.ClientParametersAuthentication;
+import com.google.api.client.auth.oauth2.RefreshTokenRequest;
+import com.google.api.client.auth.oauth2.TokenRequest;
+import com.google.api.client.auth.oauth2.TokenResponse;
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.cloud.hadoop.gcsio.GoogleCloudStorageOptions;
+import com.google.cloud.hadoop.util.AccessTokenProvider;
+import com.google.cloud.hadoop.util.HttpTransportFactory;
+import com.google.cloud.hadoop.util.RedactedString;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.flogger.GoogleLogger;
+import java.io.IOException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Optional;
+import org.apache.hadoop.conf.Configuration;
+
+public class RefreshTokenAuth2Provider implements AccessTokenProvider {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+
+  private static final AccessToken EXPIRED_TOKEN = new AccessToken("", -1L);
```

Comment:
The purpose of this magic value seems to be so that if a client calls `getAccessToken()` without first calling `refresh()`, then they should get an invalid response.

I don't think our AccessTokenProvider contract actually requires that.

I'd prefer if the call to `getAccessToken()` automatically invoked `refresh()` if it had not been done yet.

We could do that by making the `accessToken` field be an Optional<AccessToken> instead of just `AccessToken`, and then calling `refresh()` in the `getAccessToken` method if it does not yet have a value.